File: backend/app/schemas/revenue.py
from pydantic import BaseModel
from typing import List, Optional


class RevenueTrendPoint(BaseModel):
    period: str
    total_revenue: float
    new_revenue: float
    churned_revenue: float
    net_revenue: float


class RevenueTrendsResponse(BaseModel):
    period_type: str
    data: List[RevenueTrendPoint]


class SubscriptionMetricsResponse(BaseModel):
    mrr: float
    arr: float
    growth_rate: float
    churn_rate: float
    active_subscribers: int
    new_subscribers: int
    churned_subscribers: int


class ArpuResponse(BaseModel):
    overall_arpu: float
    arpu_by_plan: dict
    total_users: int
    paying_users: int


class LtvResponse(BaseModel):
    overall_ltv: float
    ltv_by_plan: dict
    avg_subscription_months: float
    total_lifetime_revenue: float


class PaymentFailureItem(BaseModel):
    transaction_id: int
    user_id: int
    amount: float
    failure_reason: str
    processed_at: str


class PaymentFailuresResponse(BaseModel):
    total_failures: int
    total_failed_amount: float
    recovery_rate: float
    recent_failures: List[PaymentFailureItem]


class ForecastPoint(BaseModel):
    period: str
    projected_revenue: float
    confidence_lower: float
    confidence_upper: float


class RevenueForecastResponse(BaseModel):
    current_mrr: float
    projected_mrr_3m: float
    projected_mrr_6m: float
    projected_mrr_12m: float
    growth_assumption: float
    forecast: List[ForecastPoint]

File: frontend/src/api/revenue.ts
import { fetchApi } from './auth';

export interface RevenueTrend {
  period: string;
  total_revenue: number;
  new_revenue: number;
  churned_revenue: number;
  net_revenue: number;
}

export interface SubscriptionMetrics {
  mrr: number;
  arr: number;
  growth_rate: number;
  churn_rate: number;
  active_subscribers: number;
  new_subscribers: number;
  churned_subscribers: number;
}

export async function getRevenueTrends(periodType = 'monthly', periods = 12) {
  return fetchApi<{ period_type: string; data: RevenueTrend[] }>(
    `/admin/analytics/revenue/trends?period_type=${periodType}&periods=${periods}`
  );
}

export async function getSubscriptionMetrics() {
  return fetchApi<SubscriptionMetrics>('/admin/analytics/revenue/metrics');
}

export async function getArpu() {
  return fetchApi<{ overall_arpu: number; arpu_by_plan: Record<string, number>; total_users: number; paying_users: number }>(
    '/admin/analytics/revenue/arpu'
  );
}

export async function getLtv() {
  return fetchApi<{ overall_ltv: number; ltv_by_plan: Record<string, number>; avg_subscription_months: number; total_lifetime_revenue: number }>(
    '/admin/analytics/revenue/ltv'
  );
}

export async function getPaymentFailures() {
  return fetchApi<{ total_failures: number; total_failed_amount: number; recovery_rate: number }>(
    '/admin/analytics/revenue/failures'
  );
}

export async function getRevenueForecast() {
  return fetchApi<{ current_mrr: number; projected_mrr_3m: number; projected_mrr_6m: number; projected_mrr_12m: number; growth_assumption: number }>(
    '/admin/analytics/revenue/forecast'
  );
}
